src/placeholder/parameter/parameter.py

```diff
@@ -0,0 +1,298 @@
+import numpy as np
+from copy import deepcopy
+
+
+from placeholder.parameter.prior import Prior
+from placeholder.exceptions import PlaceholderError
+
+BANNED_NAMES = ['intercept']
+
+
+class VariableError(PlaceholderError):
+    pass
+
+
+class ParameterError(PlaceholderError):
+    pass
+
+
+class ParameterSetError(PlaceholderError):
+    pass
+
+
+@dataclass
+class Variable:
+    """A class that stores information about a variable.
+
+    Parameters
+    ----------
+    covariate: str
+        name of the covariate for this variable. 
+    var_link_fun: callable
+        link function for this variable. 
+    fe_init: float
+        initial value to be used in optimization for fixed effect. 
+    re_init: float
+        initial value to be used in optimization for random effect.
+    re_zero_sum_std: float, optional
+        standard deviation of zero sum prior for random effects.
+    fe_gprior: Prior, optional
+        a prior of class :class:`~placeholder.parameter.prior.Prior`
+    re_gprior: Prior, optional
+        a prior of class :class:`~placeholder.parameter.prior.Prior`
+
+    Attributes
+    ----------
+    All parameters become attributes after validation.
+    
+    """
+    var_link_fun: Callable = lambda x: x
+    covariate: str = None
+
+    fe_init: float = 0.
+    re_init: float = 0.
+
+    fe_gprior: Prior = Prior()
+    re_gprior: Prior = Prior()
+
+    re_zero_sum_std: float = field(default=np.inf)
+
+    def __post_init__(self):
+        if not isinstance(self.covariate, str):
+            raise VariableError("Covariate name must be a string.")
+        if self.covariate in BANNED_NAMES:
+            raise VariableError("Choose a different covariate name that is"
+                                f"not in {BANNED_NAMES}.")
+
+
+@dataclass
+class Intercept(Variable):
+    covariate = 'intercept'
+
+
+@dataclass
+class Parameter:
+    """A class for parameters. 
+    
+    Parameters
+    ----------
+    param_name: str
+        name of the parameter 
+    link_fun: callable
+        link function for the parameter 
+    variables: List[:class:`~placeholder.parameter.parameter.Variable`]
+        a list of variables
+    
+    Attributes
+    ----------
+    All attributes from :class:`~placeholder.parameter.parameter.Variable`s in `variables` 
+    are carried over but are put into a list.
+
+    num_fe: int
+        total number of effects (variables) for the parameter
+
+    """
+    param_name: str
+    link_fun: Callable
+    variables: InitVar[List[Variable]]
+
+    num_fe: int = field(init=False)
+    covariate: List[str] = field(init=False)
+    var_link_fun: List[Callable] = field(init=False)
+
+    fe_init: List[float] = field(init=False)
+    re_init: List[float] = field(init=False)
+
+    fe_gprior: List[List[Prior]] = field(init=False)
+    re_gprior: List[List[Prior]] = field(init=False)
+    fe_bounds: List[List[Prior]] = field(init=False)
+    re_bounds: List[List[Prior]] = field(init=False)
+
```

Comment:
should we remove the bounds here too? since `Prior` has bounds in it?

Reply:
Yes! 0814ac9